File: 19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第 N 个结点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        layer = 0
        temp1 = temp2 = ListNode(0,head)
        # 先遍历拿到长度计算出删除的位置
        while head:
            layer += 1
            head = head.next
        #delete_index = layer - n 计算出需要删除的元素索引
        for _ in range(0,layer-n):    # 但是删除该元素需要到它的前一个元素，修改它的前一个元素的 next 指向即可
            temp1 = temp1.next
        # 执行 next 指向，直接跳过删除元素即可
        temp1.next = temp1.next.next
        return temp2.next
# @lc code=end

File: 19.删除链表的倒数第-n-个结点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getLength(head *ListNode) (length int) {
	for ; head != nil; head = head.Next {
		length++
	}
	return
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := getLength(head)
	temp := &ListNode{0, head}      // 返回节点的头节点
	cur := temp                     // 实际操作链表
	for i := 0; i < length-n; i++ { // length-n 不能到达，因为实际操作的节点是删除节点的上一个节点
		cur = cur.Next
	}
	cur.Next = cur.Next.Next
	return temp.Next
}

// @lc code=end

